src/libappimage/libappimage.cpp

```diff
@@ -0,0 +1,364 @@
+
+        std::vector<char> data;
+
+        XdgUtils::DesktopEntry::DesktopEntry entry;
+        // Load Desktop Entry
+        for (auto itr = appImage.files(); itr != itr.end(); ++itr) {
+            const auto& entryPath = *itr;
+            if (entryPath.find(".desktop") != std::string::npos && entryPath.find("/") == std::string::npos) {
+                itr.read() >> entry;
+                break;
+            }
+        }
+
+        auto terminalEntryValue = entry.get("Desktop Entry/Terminal");
+
+        boost::to_lower(terminalEntryValue);
+        boost::algorithm::trim(terminalEntryValue);
+
+        return terminalEntryValue == "true";
+    } catch (const std::runtime_error& err) {
+        std::clog << "Error at " << __FUNCTION__ << " : " << err.what() << std::endl;
+    } catch (...) {
+        std::clog << "Error at " << __FUNCTION__ << " : that's all we know." << std::endl;
+    }
+    return -1;
+}
+
+
+/* Return the md5 hash constructed according to
+ * https://specifications.freedesktop.org/thumbnail-spec/thumbnail-spec-latest.html#THUMBSAVE
+ * This can be used to identify files that are related to a given AppImage at a given location */
+char* appimage_get_md5(const char* path) {
+    using namespace appimage::utils;
+
+    if (path == nullptr)
+        return nullptr;
+
+    try {
+        const auto& canonicalPath = bf::weakly_canonical(path);
+
+        if (canonicalPath.empty())
+            return nullptr;
+
+        std::string uri = "file://" + UrlEncoder::encode(canonicalPath.string());
+
+        auto md5raw = HashLib::md5(uri);
+        auto md5Str = HashLib::toHex(md5raw);
+
+        return strdup(md5Str.c_str());
+    } catch (const std::runtime_error& err) {
+        std::clog << "Error at " << __FUNCTION__ << " : " << err.what() << std::endl;
+    } catch (...) {
+        std::clog << "Error at " << __FUNCTION__ << " : that's all we know." << std::endl;
+    }
+    return nullptr;
+}
+
+#ifdef LIBAPPIMAGE_DESKTOP_INTEGRATION_ENABLED
+using namespace appimage::desktop_integration;
+
+/*
+ * Register an AppImage in the system
+ * Returns 0 on success, non-0 otherwise.
+ */
+int appimage_register_in_system(const char* path, bool verbose) {
+    try {
+        AppImage appImage(path);
+        IntegrationManager manager;
+        manager.registerAppImage(appImage);
+        manager.generateThumbnails(appImage);
+
+        return 0;
+    } catch (const std::runtime_error& err) {
+        if (verbose)
+            std::clog << "Error at " << __FUNCTION__ << " : " << err.what() << std::endl;
+    } catch (...) {
+        if (verbose)
+            std::clog << "Error at " << __FUNCTION__ << " : that's all we know." << std::endl;
+    }
+    return 1;
+}
+
+
+/* Unregister an AppImage in the system */
+int appimage_unregister_in_system(const char* path, bool verbose) {
+    try {
+        AppImage appImage(path);
+        IntegrationManager manager;
+        manager.unregisterAppImage(appImage);
```

Comment:
A manager should manage a single object. That object shouldn't have to be passed on every call to it. There's no point in making instances otherwise, because why should said manager have any state? This is just weird to use. Please pass in constructor (store a const reference for instance).

Reply:
The `IntegrationManager` contains the settings that will be used in the integration/removal functions. Those settings are used in every method call, so you can use the same settings to integrate several AppImages as appimaged will do.

Reply:
A very un-obvious design. You can just apply the same settings in a for loop for that, but there's no reason not to create an object for every AppImage. I don't see any change of behavior either.

If you want some "easy" way to copy the settings, you should rather introduce a factory method that will create a new instance with the settings of the original object. But a manager should manage _one_ object. Right now, it has poorly defined responsibilities (as in, too much functionality at once).

You like Qt so much, don't you? There's a reason there's no `QFileSystemManager` that performs every operation on paths passed on every call. The _only_ methods taking a path as parameters are defined statically. As soon as instances are being used, the instance manages a single item only (e.g., a dir or file). That's much cleaner.

Reply:
This class acts a proxy over the real implementations. It also provides that factory method just that it doesn't return the actual objects created. Those objects are ("Integrator" and "Thumbnailer") are treated as functors. This abstracts our users from the integration, removal and thumbnails generation processes.

Reply:
A proxy class must implement the same interface as the real object. This can't be called a proxy therefore.

> This abstracts our users from the integration, removal and thumbnails generation processes.

A manager can manage just a single AppImage and still make these guarantees.

Reply:
Sorry, the argument is. There is no point in having such an object. Besides the AppImage it will not differ in anything from others.

Reply:
Then, there is no point in having a class either. You can just use a couple of static methods in a global namespace.

But that's _so_ annoying. The reason we have OOP is we want objects to describe real things. Things like AppImages. And perform operations on them. You're basically constructing something like "OOP in C". It's an awful pattern you're trying to 

What if you ever want to run multiple operations?

Reply:
This class is "thread safe" and reentrant as all of it's methods are const (will add that).
The point of having it is to provide a holder for the common settings to be used in the desktop integration methods without polluting the global namespace.

Reply:
Static methods can be defined just fine in any namespace. You don't have to pollute anything.